ddsketch/ddsketch.py

```diff
@@ -12,158 +45,207 @@
 
 DEFAULT_REL_ACC = 0.01  # "alpha" in the paper
 DEFAULT_BIN_LIMIT = 2048
-DEFAULT_MIN_VALUE = 1.0e-9
 
 
 class UnequalSketchParametersException(Exception):
-    pass
+    """thrown when trying to merge two sketches with different relative_accuracy
+    parameters
+    """
+
+
+class BaseDDSketch:
+    """The base implementation of DDSketch with no storage specified.
+
+    Args:
+        store (store.Store): storage for positive values
+        negative_store (store.Store): storage for negative values
+        relative_accuracty (float): the accuracy guarantee; referred to as alpha
+            in the paper. (0. < alpha < 1.)
 
+    Attributes:
+        zero_count (int): The count of zero values
+        gamma (float): the base for the exponenntial buckets
+            gamma = (1 + alpha) / (1 - alpha)
+        multiplier (float): used for calculating log_gamma(value)
+            multiplier = 1 / log(gamma)
+        _min_possible: the smallest value the sketch can distinguish from 0
+
+        count: the number of values seen by the sketch
+        min: the minimum value seen by the sketch
+        max: the maximum value seen by the sketch
+        sum: the sum of the values seen by the sketch
+    """
 
-class BaseDDSketch(object):
     def __init__(
-        self, relative_accuracy=None, bin_limit=None, min_value=None, store=None
+        self,
+        store,
+        negative_store,
+        relative_accuracy,
     ):
-        # Make sure the parameters are valid
-        if relative_accuracy is None or (
-            relative_accuracy <= 0 or relative_accuracy >= 1
-        ):
-            relative_accuracy = DEFAULT_REL_ACC
-        if bin_limit is None or bin_limit < 0:
-            bin_limit = DEFAULT_BIN_LIMIT
-        if min_value is None or min_value < 0:
-            min_value = DEFAULT_MIN_VALUE
-        if store is None:
-            self.store = CollapsingLowestDenseStore(bin_limit)
-        else:
-            self.store = store
+        self.store = store
+        self.negative_store = negative_store
+        self.relative_accuracy = relative_accuracy
+
+        self.zero_count = 0
 
-        x = 2 * relative_accuracy / (1 - relative_accuracy)
-        self.gamma = 1 + x
-        self.gamma_ln = math.log1p(x)
-        self.min_value = min_value
-        self.offset = -int(math.ceil(math.log(min_value) / self.gamma_ln)) + 1
+        gamma_mantissa = 2 * relative_accuracy / (1 - relative_accuracy)
+        self.gamma = 1 + gamma_mantissa
+        gamma_ln = math.log1p(gamma_mantissa)
+        self.multiplier = 1.0 / gamma_ln
+        self._min_possible = np.finfo(np.float64).tiny * self.gamma
 
-        self._min = float("+inf")
-        self._max = float("-inf")
-        self._count = 0
-        self._sum = 0
+        self.count = 0
+        self.min = float("+inf")
+        self.max = float("-inf")
+        self.sum = 0.0
```

Comment:
This very well could be my fault, but I pulled this branch locally to debug something with my own implementation, and I think this is breaking on initializing a DDSketch without parameters? I.e.

```py
sketch = DDSketch()
```

Log output:
```sh
sketches-py git:homin/negative_store ❯ python3

>>> from ddsketch.ddsketch import DDSketch
>>> sketch = DDSketch()

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/tim.brown/go/src/github.com/DataDog/sketches-py/ddsketch/ddsketch.py", line 247, in __init__
    super().__init__(
  File "/Users/tim.brown/go/src/github.com/DataDog/sketches-py/ddsketch/ddsketch.py", line 100, in __init__
    self.sum = 0.0
AttributeError: can't set attribute
```


Does `sum` need to be treated differently in the constructor because it's defined with `@property`?

Reply:
derp; there was a reason why it was `_sum` before I undid it. 🤦